### PR TITLE
AI Guidelines file

### PR DESCRIPTION
**Pull request by kaapstorm**
## Technical Summary

This PR adds a guidelines file for AI Assistants.

It was inspired by `CLAUDE.md` in the open-chat-studio repo. It was written in collaboration with Junie. Junie's suggestions and feedback turned out to be a lot less specific than the content of open-chat-studio's `CLAUDE.md`. I found that interesting, but I kept Junie's more general approach. (Also interesting was that Junie repeatedly suggested content intended for the developer, which I have kept.)

The PR also adds symlinks to the file for Claude and Junie. ( :blowfish: Review by commit.)

Cursor uses multiple rules files, for different contexts (e.g. rules for Python, rules for JavaScript, etc.) so this file would not be appropriate for Cursor.

I am keen to hear feedback on whether this file is useful to other developers.

I am cautiously curious to hear feedback on the content of the file. (I have found that my Python style is in some cases more conservative than other HQ devs, but that might do well for an AI Assistant because the style is the least likely to offend. My approach to commits by AI agents is also more conservative than some developers.)

### Safety story

A documentation-only change.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: ai_guidelines.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+### For the AI Assistant
+
+- The AI Assistant should maintain professional, focused communication.
+- The AI Assistant should ask clarifying questions whenever details are unclear,
+  or choices are not obvious.
+- When appropriate, the AI Assistant should make recommendations on how the
+  Developer can improve their prompt to get a more effective response.
+
+### For the Developer
```
Comment (snopoke) [inline ai_guidelines.md]: I don't think this is valuable to have here

File: ai_guidelines.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+## General Principles
+
+- Focus on clarity, maintainability, and performance
+- Consider both immediate needs and long-term implications
+
+## Communication
+
+### For the AI Assistant
+
+- The AI Assistant should maintain professional, focused communication.
```
Comment (snopoke) [inline ai_guidelines.md]: This doesn't seem useful.

File: ai_guidelines.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+- The Developer should provide feedback to the AI Assistant to improve future
+  responses.
+- The Developer should update this document to assist the AI Assistant across
+  sessions, and to improve the interactions of other Developers.
+
+## Tech Stack
+
+- Backend: Python 3.9, Django 4.2
+- Frontend: JavaScript (ECMAScript 2015 (ES6)), HTMX, Alpine.js, Knockout.js
```
Comment (snopoke) [inline ai_guidelines.md]: what about css frameworks and blundlers

File: ai_guidelines.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  (legacy)
+- Databases: PostgreSQL, Elasticsearch, CouchDB (legacy)
+- Asynchronous task queue: Celery
+- Cache & message broker: Redis
+- Stream processing: Kafka
+- Version Control: Git
```
Comment (snopoke) [inline ai_guidelines.md:38]: It's also useful to add linting tools

File: ai_guidelines.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+### For the Developer
+
+- The Developer should provide feedback to the AI Assistant to improve future
+  responses.
+- The Developer should update this document to assist the AI Assistant across
+  sessions, and to improve the interactions of other Developers.
+
+## Tech Stack
+
+- Backend: Python 3.9, Django 4.2
```
Comment (millerdev) [inline ai_guidelines.md]: It presents a maintenance burden to have hard-coded versions here. Would it work to instead reference pyproject.toml and uv.lock as the source of truth for version information?

Or maybe just remove the versions? None of the components below except for JavaScript have version numbers.

Comment (kaapstorm) [reply ai_guidelines.md]: 862762e4a0f

File: ai_guidelines.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+- Comments should explain "why" rather than "what": As much as possible,
+  "what" should be clear from the code.
+- Keep module README files and project documentation under `docs/`
+  up-to-date.
+
+### Security
+
+- Are there any security vulnerabilities?
+- Is user input properly validated and sanitized?
+- Are permissions and access controls properly implemented?
```
Comment (snopoke) [inline ai_guidelines.md]: I'm not sure about having questions in this. It makes sense for code review but less so for code generation. I expect code review would be fine with affirmative statements too e.g. Be mindful of security and follow best practices for handling user input.

File: ai_guidelines.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+### Testing
+
+- Are there appropriate tests for the changes?
+- Do the tests cover edge cases and failure scenarios?
+- Are the tests clear and maintainable?
+- Are doctests tested?
+- Run the test suite before submitting changes
```
Comment (snopoke) [inline ai_guidelines.md]: Is this practical for HQ's tests?

File: ai_guidelines.md (hunk 2 above)
Comment (kaapstorm) [reply ai_guidelines.md]: In version 1, this said "The AI should not flatter the Developer" (after several cases of things like "That's a great idea" and "Good choice!"). Junie's feedback on that was that this sentence would cover that, and so I went with this instead.

Have you noticed flattery? If so, do you like it, or doesn't it bother you? I'm okay with removing this. But I'm pretty certain that not all my choices are genuinely good ones! :joy:

File: ai_guidelines.md (hunk 1 above)
Comment (kaapstorm) [reply ai_guidelines.md]: Yeah, I thought so too, and I removed Junie's initial section on guidelines for the developer. But then Junie suggested similar content again, so I added this. I'm happy to remove it. :thinking: I should dig a little deeper though, and ask Junie what the intention is with these repeated suggestions, considering the purpose and audience of this document.

File: ai_guidelines.md (hunk 7 above)
Comment (kaapstorm) [reply ai_guidelines.md]: With Junie and Claude, I tend to tell it how to run the tests that cover the area I'm working on.

File: ai_guidelines.md (hunk 2 above)
Comment (snopoke) [reply ai_guidelines.md]: I haven't noticed anything in the tools I've used.

File: ai_guidelines.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+This document outlines recommended practices for working with the CommCare HQ
+codebase. These guidelines apply in the following circumstances:
+
+- Reviewing code or suggesting changes
+- Writing code and making changes
+- Assisting with version control
+
+## General Principles
+
+- Focus on clarity, maintainability, and performance
```
Comment (mjriley) [inline ai_guidelines.md]: This line, as phrased, doesn't seem that useful. Wouldn't an AI just read this as "write good code that isn't overly terse"? I think the interesting part is more about when clarity, maintainability, and performance come into conflict. I'd prefer if this was worded to say to focus on clarity and maintainability, with a lesser focus on performance. Situations where necessary performance comes at the cost of either clarity or maintainability should be documented through comments

File: ai_guidelines.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+- Consider potential performance bottlenecks.
+- For database operations, queries should be optimized. In rare instances this
+  could require the Developer to test the performance of the query in an
+  environment similar to production. If the AI Assistant is writing code, then
+  notify the Developer when this would be beneficial.
+- For front-end code, be mindful of rendering or loading concerns.
+
+### Maintainability
+
+- Have comments or docstrings that are affected by the change been updated?
```
Comment (mjriley) [inline ai_guidelines.md]: As the other guidelines seem to be declarative, would this be better phrased as "Update comments and docstrings affected by incoming changes"?

File: ai_guidelines.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+- Comments should explain "why" rather than "what": As much as possible,
+  "what" should be clear from the code.
+- Keep module README files and project documentation under `docs/`
+  up-to-date.
+
+### Security
+
+- Always consider security vulnerabilities.
+- Ensure that user input is properly validated and sanitized.
+- Ensure that permissions and access controls are properly implemented.
```
Comment (mjriley) [inline ai_guidelines.md:80]: Do we have any informal requirements about trying to make destructive things less permanent? Thinking of how auditing requirements influence how we disable rather than delete things like domains. I wonder if that would be helpful for an AI

File: ai_guidelines.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+### Testing
+
+- Changes must be covered by appropriate tests.
+- Tests need to cover edge cases and failure scenarios.
+- Tests should be clear and maintainable.
+- Doctests should be tested.
+- Run the tests that cover the changes before considering any changes complete.
+- Use pytest features and pytest conventions, like Pythonic `assert`
```
Comment (mjriley) [inline ai_guidelines.md:92]: Not relevant to this review, but I'm curious if we've circulated this guidance to the team. I wasn't aware of this recommendation

File: ai_guidelines.md (hunk 8 above)
Comment (biyeun) [reply ai_guidelines.md]: agree with matt here. sometimes the tendency with AI is to prematurely optimize when asked to focus on performance. We want to emphasize clarity, maintainablity first. We can prompt specifically for performance later when there are issues. Might also specifically ask to write DRY code whenever possible as well.

File: ai_guidelines.md [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+## Writing and Reviewing Code
+
+### Code Clarity
+
+- Code should be readable and well-structured.
+- Code should follow the conventions used in the rest of the module.
+- When writing or reviewing code, consider any potential bugs or edge cases.
+- Keep functions and methods focused on a single responsibility. Break up
+  longer functions where appropriate.
```
Comment (biyeun) [inline ai_guidelines.md:48]: here i would add, "Keep code DRY whenever possible" (or something like this)

File: ai_guidelines.md [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+- When appropriate, the AI Assistant should make recommendations on how the
+  Developer can improve their prompt to get a more effective response.
+
+## Tech Stack
+
+- Backend: Python, Django
+- Python dependency management: uv
+- Testing: pytest
+- Linting & formatting: Flake8, isort, YAPF
+- Frontend: JavaScript, HTMX, Alpine.js, Knockout.js (legacy), Bootstrap
```
Comment (biyeun) [inline ai_guidelines.md]: should we be specific that some pages are still on Bootstrap 3 and we are in the process of migrating from Bootstrap 3 to 5?

File: ai_guidelines.md (hunk 11 above)
Comment (kaapstorm) [reply ai_guidelines.md:92]: @millerdev someone else has said this to me before. Is this guideline correct, and if so, is it worth mentioning on `#gtd-web-team`?

Comment (millerdev) [reply ai_guidelines.md:92]: I think it's mostly correct. One notable exception is magical pytest fixtures, which are not allowed in HQ code; [pytest-unmagic](https://github.com/dimagi/pytest-unmagic/) should be used instead.

Comment (kaapstorm) [reply ai_guidelines.md:92]: > [pytest-unmagic](https://github.com/dimagi/pytest-unmagic/) should be used instead

:+1:  That is stated [on the next line](https://github.com/dimagi/commcare-hq/blob/3373daec2ca656367eb1a4547ad0cc0c67b9fc23/ai_guidelines.md?plain=1#L94).

**Review by biyeun — APPROVED** [review]
I'll give this one a ✅ @kaapstorm 
sorry for the big delay
